src/ResourceManager/Kubernetes/Commands.Kubernetes/Commands/CreateOrUpdateKubeBase.cs

```diff
@@ -0,0 +1,363 @@
+using Microsoft.Azure.Management.Internal.Resources;
+using Microsoft.Rest.Azure;
+using Newtonsoft.Json;
+
+namespace Microsoft.Azure.Commands.Kubernetes
+{
+    [Cmdlet(VerbsCommon.New, KubeNounStr, DefaultParameterSetName = DefaultParamSet)]
+    [OutputType(typeof(PSObject), typeof(List<PSObject>))]
+    public abstract class CreateOrUpdateKubeBase : KubeCmdletBase
+    {
+        protected const string DefaultParamSet = "defaultParameterSet";
+        protected const string SpParamSet = "servicePrincipalParameterSet";
+        protected readonly Regex _dnsRegex = new Regex("[^A-Za-z0-9-]");
+
+        [Parameter(
+            Position = 0,
+            Mandatory = true,
+            ParameterSetName = DefaultParamSet,
+            HelpMessage = "Resource Group Name.")]
+        [Parameter(
+            Position = 0,
+            Mandatory = true,
+            ParameterSetName = SpParamSet,
+            HelpMessage = "Resource Group Name.")]
+        [ResourceGroupCompleter()]
+        [ValidateNotNullOrEmpty]
+        public string ResourceGroupName { get; set; }
+
+        [Parameter(
+            Position = 1,
+            Mandatory = true,
+            ParameterSetName = DefaultParamSet,
+            HelpMessage = "Kubernetes managed cluster Name.")]
+        [Parameter(
+            Position = 1,
+            Mandatory = true,
+            ParameterSetName = SpParamSet,
+            HelpMessage = "Kubernetes managed cluster Name.")]
+        [ValidateNotNullOrEmpty]
+        [ValidatePattern("^[a-zA-Z0-9][a-zA-Z0-9_.-]*$")]
+        [ValidateLength(2, 64)]
+        public string Name { get; set; }
+
+        [Parameter(
+            Position = 2,
+            Mandatory = true,
+            ParameterSetName = SpParamSet,
+            HelpMessage =
+                "The client ID of the AAD application / service principal used for cluster authentication to Azure APIs.")]
+        public string ClientId { get; set; }
+
+        [Parameter(
+            Position = 3,
+            Mandatory = true,
+            ParameterSetName = SpParamSet,
+            HelpMessage = "The secret associated with the AAD application / service principal.")]
+        public string ClientSecret { get; set; }
+
+        [Parameter(Mandatory = false,
+            HelpMessage = "Azure location for the cluster. Defaults to the location of the resource group.")]
+        [LocationCompleter("Microsoft.ContainerService/managedClusters")]
+        public string Location { get; set; }
+
+        [Parameter(Mandatory = false, HelpMessage = "User name for the Linux Virtual Machines.")]
+        public string AdminUserName { get; set; } = "azureuser";
+
+        [Parameter(Mandatory = false, HelpMessage = "The DNS name prefix for the cluster.")]
+        public string DnsNamePrefix { get; set; }
+
+
+        [Parameter(Mandatory = false, HelpMessage = "The version of Kubernetes to use for creating the cluster.")]
+        [ValidateSet("1.7.7", "1.8.1")]
```

Comment:
I am assuming that this will change over time - if so, this shoudl use a GenericArgumentCompleter, rather than a ValidateSet